File: src/erpl/engine/model/precondition_tree.py
class PreConditionTree():
    def __init__(self,root):
        self.root = root

    def test_tree(self,room, inventory,state):
        return self.root.test_node(room, inventory,state)

class PreConditionNode():
    def __init__(self,value,left,right, is_operator=False):
        self.value = value
        self.left = left
        self.right = right
        self.is_operator = is_operator

    def test_node(self, room, inventory,state):
        if self.is_operator:
            if self.value == "and":
                return self.left.test_node(room, inventory,state) and self.right.test_node(room, inventory,state)
            elif self.value == "or":
                return self.left.test_node(room, inventory,state) or self.right.test_node(room, inventory,state)
            elif self.value == "not":
                return not self.left.test_node(room, inventory,state)
        else:
            return self.value.test(room, inventory,state)
            #return testar self.value

class PreConditionOperatorAnd(PreConditionNode):
    def __init__(self, left, right):
        super().__init__("and", left, right, is_operator=True)

class PreConditionOperatorOr(PreConditionNode):
    def __init__(self, left, right):
        super().__init__("or", left, right, is_operator=True)

class PreConditionOperatorNot(PreConditionNode):
    def __init__(self, left):
        super().__init__("not", left, None, is_operator=True)

class PreConditionVar(PreConditionNode):
    def __init__(self, value):
        super().__init__(value, None, None)

File: erpl/erpl/engineP5/model/precondition_tree.js
// Classe PreConditionTree
 class PreConditionTree {
  constructor(root) {
      this.root = root;
  }

  testTree(room, inventory, state) {
      return this.root.testNode(room, inventory, state);
  }
}

// Classe PreConditionNode
 class PreConditionNode {
  constructor(value, left = null, right = null, isOperator = false) {
      this.value = value;
      this.left = left;
      this.right = right;
      this.isOperator = isOperator;
  }

  testNode(room, inventory, state) {
      if (this.isOperator) {
          if (this.value === "and") {
              return this.left.testNode(room, inventory, state) && this.right.testNode(room, inventory, state);
          } else if (this.value === "or") {
              return this.left.testNode(room, inventory, state) || this.right.testNode(room, inventory, state);
          } else if (this.value === "not") {
              return !this.left.testNode(room, inventory, state);
          }
      } else {
          return this.value.test(room, inventory, state);
      }
  }
}

// Classes para Operadores Lógicos
 class PreConditionOperatorAnd extends PreConditionNode {
  constructor(left, right) {
      super("and", left, right, true);
  }
}

 class PreConditionOperatorOr extends PreConditionNode {
  constructor(left, right) {
      super("or", left, right, true);
  }
}

 class PreConditionOperatorNot extends PreConditionNode {
  constructor(left) {
      super("not", left, null, true);
  }
}

// Classe PreConditionVar
 class PreConditionVar extends PreConditionNode {
  constructor(value) {
      super(value);
  }
}
